### PR TITLE
fix: Slack report shows duplicate findings per PR in multi-project mode

### LINKED ISSUE #195: fix: Slack report shows duplicate findings per PR in multi-project mode
## Problem

In multi-project mode, the Slack report shows the same findings repeated 2-3 times per PR:

```
openshift/hypershift (2 PR(s)):
  PR #8365 — OCPBUGS-77307: Generate KubeVirt nmstate network config conditionally
    :warning: PR #8365 is behind main
    :speech_balloon: PR #8365 has 1 new review comment(s)
    :warning: PR #8365 is behind main
    :speech_balloon: PR #8365 has 1 new review comment(s)
    :warning: PR #8365 is behind main
    :speech_balloon: PR #8365 has 1 new review comment(s)
```

## Root Cause

In multi-project mode (`runMultiProject` in `cmd/oompa/main.go`), there are two unsynchronized timers:

1. Each role goroutine polls every `poll-interval` (e.g., 30s) and calls `CollectSlackFindings()` which **appends** findings to the shared `SlackReporter.pending` slice.
2. A separate flush goroutine fires on its own `poll-interval` ticker and calls `Flush()` which drains `pending` and sends to Slack.

Because the timers start at different times and aren't synchronized, 2-3 poll cycles can accumulate the **same findings** in `pending` before the flush fires:

```
t=0s   Poll cycle 1: Collect("rebase:8365", "review:8365")  → pending has 2
t=30s  Poll cycle 2: Collect("rebase:8365", "review:8365")  → pending has 4
t=60s  Poll cycle 3: Collect("rebase:8365", "review:8365")  → pending has 6
t=~60s Flush fires: all 6 pass dedup (first flush), 3x duplicates in Slack
```

The dedup in `Flush()` (lines 123-129 of `slack.go`) only checks against `r.reported` -- findings from **previously flushed** batches. It does not check for duplicates **within the current batch**.

## Fix

Deduplicate within the pending batch at flush time, before checking against `r.reported`. Add this in `Flush()` in `pkg/agent/slack.go`, after draining `pending` (after line 111) and before the `r.reported` dedup check:

```go
// Deduplicate within the pending batch (multiple poll cycles may have
// collected the same finding before this Flush ran).
seen := make(map[string]bool)
var deduped []SlackFinding
for _, f := range findings {
    if f.DedupKey != "" {
        if seen[f.DedupKey] {
            continue
        }
        seen[f.DedupKey] = true
    }
    deduped = append(deduped, f)
}
findings = deduped
```

This is preferred over deduplicating at `Collect()` time because:
- It's a single localized change in `Flush()`
- `Collect()` stays as a simple O(1) append (important for concurrent callers)
- It handles edge cases where different role goroutines collect the same finding

## Tests

1. Multiple poll cycles collecting the same findings → single entry in Slack message
2. Different findings for the same PR → all appear (no over-dedup)
3. Findings without a DedupKey → always included
4. Same finding across two different flush cycles → deduped by `r.reported` (existing behavior, unchanged)

## Files to change

| File | Change |
|------|--------|
| `pkg/agent/slack.go` | Add within-batch dedup in `Flush()` after draining `pending` |
| `pkg/agent/slack_test.go` | Add test for duplicate findings within a single flush batch |

### PR DESCRIPTION
**Pull request by qinqon**
Fixes #195

## Summary

Fix duplicate findings in Slack reports when running in multi-project mode. Multiple poll cycles could append identical findings to the pending buffer before Flush ran, causing the same finding to appear 2-3 times per PR in the Slack message.

## Changes

- Add within-batch deduplication in `SlackReporter.Flush()` after draining `pending` and before the existing `r.reported` cross-flush dedup check. Findings with the same `DedupKey` within a single batch are collapsed to one entry; findings with empty `DedupKey` are never deduplicated.
- Add three new tests covering the within-batch dedup scenarios: duplicate findings from multiple poll cycles, different findings for the same PR (no over-dedup), and empty `DedupKey` findings always included.

## Testing

- [x] `go test ./...` passes
- [x] `go vet ./...` passes
- [ ] Manual testing (if applicable)

## Related Issues

Fixes #195

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved deduplication of findings in Slack reports to prevent duplicate notifications when identical findings are collected within a single batch.

[Review Change Stack](https://app.coderabbit.ai/change-stack/qinqon/oompa/pull/196?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: pkg/agent/slack.go [hunk 1]
```diff
@@ -114,6 +114,21 @@ func (r *SlackReporter) Flush(ctx context.Context) {
 		return
 	}
 
+	// Deduplicate within the pending batch (multiple poll cycles may have
+	// collected the same finding before this Flush ran).
+	seen := make(map[string]bool)
+	var deduped []SlackFinding
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go]: > To avoid multiple slice reallocations as findings are appended, pre-allocate the `deduped` slice with the maximum possible capacity (`len(findings)`).

Addressed: Pre-allocated `deduped` with `make([]SlackFinding, 0, len(findings))`.

File: pkg/agent/slack_test.go [hunk 2]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+
+func TestSlackReporter_FlushDedup_WithinBatch(t *testing.T) {
+	var postCount int
+	var receivedBody string
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		postCount++
+		body, _ := io.ReadAll(r.Body)
+		receivedBody = string(body)
+		w.WriteHeader(http.StatusOK)
+	}))
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:829]: > Accessing `postCount` and `receivedBody` from both the handler goroutine and the main test goroutine without synchronization introduces a data race. Use a `sync.Mutex` to protect these shared variables.

Addressed: Added `sync.Mutex` to protect `postCount` and `receivedBody` in the handler, and read them under the lock in assertions.

File: pkg/agent/slack_test.go [hunk 3]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+
+	// Verify the message contains each finding exactly once
+	rebaseCount := strings.Count(receivedBody, "behind main")
+	if rebaseCount != 1 {
+		t.Errorf("expected 1 'behind main' in message, got %d", rebaseCount)
+	}
+	reviewCount := strings.Count(receivedBody, "new review comment")
+	if reviewCount != 1 {
+		t.Errorf("expected 1 'new review comment' in message, got %d", reviewCount)
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go]: > Read the shared variables under the protection of the mutex to prevent data races.

Addressed: Assertions now read `postCount` and `receivedBody` under the mutex lock.

File: pkg/agent/slack_test.go [hunk 4]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+}
+
+func TestSlackReporter_FlushDedup_DifferentFindingsSamePR(t *testing.T) {
+	var receivedBody string
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, _ := io.ReadAll(r.Body)
+		receivedBody = string(body)
+		w.WriteHeader(http.StatusOK)
+	}))
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:880]: > Accessing `receivedBody` from both the HTTP handler goroutine and the main test goroutine without synchronization introduces a data race. Use a `sync.Mutex` to protect it.

Addressed: Added `sync.Mutex` to protect `receivedBody` in the handler and read it under the lock in assertions.

File: pkg/agent/slack_test.go [hunk 5]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+
+	if !strings.Contains(receivedBody, "e2e failed") {
+		t.Error("expected 'e2e failed' in message")
+	}
+	if !strings.Contains(receivedBody, "behind main") {
+		t.Error("expected 'behind main' in message")
+	}
+	if !strings.Contains(receivedBody, "new review comments") {
+		t.Error("expected 'new review comments' in message")
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:910]: > Read the shared `receivedBody` variable under the protection of the mutex to prevent data races.

Addressed: Assertions now read `receivedBody` under the mutex lock.

File: pkg/agent/slack_test.go [hunk 6]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+}
+
+func TestSlackReporter_FlushDedup_EmptyDedupKeyNotDeduped(t *testing.T) {
+	var receivedBody string
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, _ := io.ReadAll(r.Body)
+		receivedBody = string(body)
+		w.WriteHeader(http.StatusOK)
+	}))
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:922]: > Accessing `receivedBody` from both the HTTP handler goroutine and the main test goroutine without synchronization introduces a data race. Use a `sync.Mutex` to protect it.

Addressed: Added `sync.Mutex` to protect `receivedBody` in the handler and read it under the lock in assertions.

File: pkg/agent/slack_test.go [hunk 7]
```diff
@@ -814,3 +814,113 @@ func TestFormatSlackMessage_TruncatesAt50Blocks(t *testing.T) {
...
+		})
+	}
+
+	reporter.Flush(ctx)
+
+	// All 3 should appear (empty DedupKey is never deduplicated)
+	errorCount := strings.Count(receivedBody, "transient error")
+	if errorCount != 3 {
+		t.Errorf("expected 3 'transient error' in message (empty DedupKey), got %d", errorCount)
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:948]: > Read the shared `receivedBody` variable under the protection of the mutex to prevent data races.

Addressed: Assertions now read `receivedBody` under the mutex lock.

File: pkg/agent/slack.go [hunk 8]
```diff
@@ -114,6 +114,21 @@ func (r *SlackReporter) Flush(ctx context.Context) {
...
+	for _, f := range findings {
+		if f.DedupKey != "" {
+			if seen[f.DedupKey] {
+				continue
+			}
+			seen[f.DedupKey] = true
+		}
+		deduped = append(deduped, f)
+	}
+	findings = deduped
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go:130]: > Within-batch dedup keys on `f.DedupKey` alone. In multi-project mode the reporter is shared across repos, but several DedupKey generators don't include owner/repo...

Not addressing: This is a pre-existing concern, not a regression from this PR. The within-batch dedup uses the exact same `DedupKey` field that the cross-flush dedup (`r.reported` check at line 141) already uses. If DedupKey collisions between repos were a problem, the existing `r.reported` dedup would already suppress findings incorrectly across flush cycles. Fixing DedupKey namespacing would require changing the generators in `loop.go` and both dedup paths -- that's a separate issue beyond the scope of this fix.

**Comment by qinqon** [comment]
## CI Failure Analysis — commit 0db590b

**Total failures**: 1 (1 related)
**Action taken**: Pushed fix for related failure

---

<details>
<summary>🔴 Related: <code>test</code> — fix pushed</summary>

### Error
````
```
diff --git a/pkg/agent/triage.go b/pkg/agent/triage.go
-	idx := strings.Index(text, heading)
-	if idx < 0 {
+	_, after, ok := strings.Cut(text, heading)
+	if !ok {
-	rest := text[idx+len(heading):]
+	rest := after
make: *** [Makefile:22: check-fix] Error 1
go fix ./... produced changes; run 'make fix' and commit
```
````

### Root Cause
The CI pipeline runs `make check-fix` which executes `go fix ./...` and then `git diff --exit-code` to ensure no changes are produced. The Go toolchain's `go fix` now rewrites `strings.Index`+slice patterns to use `strings.Cut`. While `triage.go` is not modified by this PR, the check runs across the entire codebase and the pre-existing pattern in `extractSection()` triggers the failure.

### Recommendation
Applied `go fix` output to `pkg/agent/triage.go` and amended into the commit. CI should now pass.

### Action
Pushed fix for this failure.

</details>

Pushed a fix for the related failure.
